src/cpp/ext/gsm/metadata_exchange.cc

```diff
@@ -120,6 +122,116 @@ absl::string_view GetStringValueFromUpbStruct(google_protobuf_Struct* struct_pb,
   return "unknown";
 }
 
+class LocalLabelsIterable : public LabelsIterable {
+ public:
+  explicit LocalLabelsIterable(
+      const std::vector<std::pair<absl::string_view, std::string>>& labels)
+      : labels_(labels) {}
+
+  absl::optional<std::pair<absl::string_view, absl::string_view>> Next()
+      override {
+    if (pos_ >= labels_.size()) {
+      return absl::nullopt;
+    }
+    return labels_[pos_++];
+  }
+
+  size_t Size() const override { return labels_.size(); }
+
+  void ResetIteratorPosition() override { pos_ = 0; }
+
+ private:
+  size_t pos_ = 0;
+  const std::vector<std::pair<absl::string_view, std::string>>& labels_;
+};
+
+class PeerLabelsIterable : public LabelsIterable {
+ public:
+  explicit PeerLabelsIterable(grpc_core::Slice remote_metadata) {
+    std::string decoded_metadata;
+    bool metadata_decoded = absl::Base64Unescape(
+        remote_metadata.as_string_view(), &decoded_metadata);
+    if (metadata_decoded) {
+      struct_pb_ = google_protobuf_Struct_parse(
+          decoded_metadata.c_str(), decoded_metadata.size(), arena_.ptr());
+      type_ = GetStringValueFromUpbStruct(struct_pb_, kMetadataExchangeTypeKey,
+                                          arena_.ptr());
+    }
+  }
+
+  absl::optional<std::pair<absl::string_view, absl::string_view>> Next()
+      override {
+    if (struct_pb_ == nullptr) {
+      return absl::nullopt;
+    }
+    if (pos_++ == 0) {
+      return std::make_pair(kPeerTypeAttribute, type_);
+    }
+    // Only handle GKE type for now.
+    if (type_ != kGkeType) {
+      return absl::nullopt;
+    }
+    switch (pos_ - 1) {
```

Comment:
Why subtract 1 here?  Seems like the case statements could just be incremented by 1.

Reply:
done